pygfx/controllers/_base.py

```diff
@@ -101,10 +184,93 @@ def register_events(self, viewport_or_renderer: Union[Viewport, Renderer]):
             "key_down",
             "key_up",
             "wheel",
+            "before_render",
         )
 
-    def handle_event(self, event, viewport, camera):
-        raise NotImplementedError()
+    def tick(self):
+        """Should be called periodically to keep the camera up to date.
+
+        If ``auto_update`` is True, this is done automatically. Returns
+        a dict with the new camera state, or None if there are no
+        running actions.
+        """
+
+        if not self._actions:
+            return None
+
+        # Get elapsed time since last frame
+        now = perf_counter()
+        elapsed_time = now - self._last_tick_time
+        self._last_tick_time = now
+        # Determine damping/smoothing factor to update action values.
+        # In the formula below the mul with elapsed_time is equivalent
+        # to a division by fps. The mul with 50 is just so typical
+        # values for damping lay between 0..10 :)
+        factor = 1
+        if self.damping > 0:
+            factor = min(1, 50 * elapsed_time / self.damping)
+
+        to_pop = []
+        for key, action in self._actions.items():
+            if action.mode == "repeat" and not action.done:
+                action.increase_target(elapsed_time)
+            action.tick(factor)
+            if action.is_at_target and action.done:
+                to_pop.append(key)
+            self._apply_action(action)
+
+        # Remove actions that are done
+        for key in to_pop:
+            self._actions.pop(key)
+
+        return self._update_cameras()
```

Comment:
I see how this returns camera state, but it seems like it still always calls `camera.set_state(...)` inside `self._update_cameras()` regardless of `auto_update`?

Reply:
Woops! I updated the docs, but forgot the code 😄